### PR TITLE
Namespacing: Datadog::Core::Buffer

### PR DESCRIPTION
**Pull request by delner**
This pull request moves some `Buffer` classes to `Datadog::Core::Buffer`.

### Related PRs

It's part of a series of `Core` namespacing PRs derived from `refactor/core-namespacing`:

1. #1847
2. #1848
3. #1850 
4. #1851
5. #1852 
6. #1853 
7. #1854 
8. #1855 
9. **=> Namespacing: Datadog::Core::Buffer**
10. #1857
11. #1858
12. #1859 
13. #1860 
14. #1861 
15. #1862

### CODE REVIEW

File: lib/datadog/core/buffer/cruby.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+# typed: true
+require 'datadog/core/buffer/random'
+
+module Datadog
+  # Buffer that accumulates items for a consumer.
+  # Consumption can happen from a different thread.
+
+  # Buffer that stores objects. The buffer has a maximum size and when
+  # the buffer is full, a random object is discarded.
+  module Core
```
Comment (ivoanjo) [inline lib/datadog/core/buffer/cruby.rb:10]: This comment should be moved/merged with `class CRuby` below, as right now it's being attached to `Datadog::Core`

(Same note applies elsewhere in the PR)

File: lib/ddtrace/buffer.rb [hunk 2]
```diff
@@ -1,229 +1,11 @@
...
-
-      # We should replace a random trace with the new one
-      replace_index = rand(@max_size)
-      @items[replace_index] = item
-    end
-  end
-
   # Health metrics for trace buffers.
   module MeasuredBuffer
     include Kernel # Ensure that kernel methods are always available (https://sorbet.org/docs/error-reference#7003)
```
Comment (ivoanjo) [inline lib/ddtrace/buffer.rb:11]: ⬇️  Just to confirm, we're leaving the rest here because it's tracing-specific (and thus it'll probably moved on a different PR inside the new tracing folder?)

File: spec/datadog/core/buffer/shared_examples.rb [hunk 3]
```diff
@@ -0,0 +1,297 @@
+# typed: false
+require 'spec_helper'
+
+require 'benchmark'
+require 'concurrent'
+
+RSpec.shared_examples 'thread-safe buffer' do
+  subject(:buffer) { described_class.new(max_size) }
```
Comment (ivoanjo) [inline spec/datadog/core/buffer/shared_examples.rb:8]: The contents of this file were moved over from `spec/ddtrace/buffer_spec.rb` but not removed from there. Can you doublecheck if this is intended?

**Review by ivoanjo — APPROVED** [review]
Left a few notes, but overall LGTM 👍

File: lib/ddtrace/buffer.rb (hunk 2 above)
Comment (delner) [reply lib/ddtrace/buffer.rb:11]: Yes, it will move to `datadog/tracing`. This is just an intermediate step.

File: spec/datadog/core/buffer/shared_examples.rb (hunk 3 above)
Comment (delner) [reply spec/datadog/core/buffer/shared_examples.rb:8]: Yes, this is intentional. Both examples in `Core` and `ddtrace` need this example group. For lack of inheriting test components from `Core`, I think this should be acceptable to duplicate for now. Something we can always optimize if we have time to tune up the test suite later.

Comment (ivoanjo) [reply spec/datadog/core/buffer/shared_examples.rb:8]: Seems reasonable. I'd only suggest calling it out with a comment on both files, so that we're aware of this duplication and the reasoning behind it.
